lib/internal/process/execution.js

```diff
@@ -238,10 +220,283 @@ function readStdin(callback) {
+      emitExperimentalWarning('Type Stripping');
+    } catch (tsError) {
+      // If its not an error, or it's not an invalid typescript syntax error, rethrow it.
+      if (!isError(tsError) || tsError?.code !== 'ERR_INVALID_TYPESCRIPT_SYNTAX') {
+        throw tsError;
+      }
+
+      try {
+        originalError.stack = decorateCJSErrorWithTSMessage(originalError.stack, tsError.message);
+      } catch { /* Ignore potential errors coming from `stack` getter/setter */ }
+      throw originalError;
+    }
+  }
+
+  if (shouldLoadESM) {
+    return require('internal/modules/run_main').runEntryPointWithESMLoader(
+      () => runScriptInContext(name,
+                               sourceToRun,
+                               breakFirstLine,
+                               print,
+                               module,
+                               baseUrl,
+                               compiledScript,
+                               origModule));
+  }
+
+  runScriptInContext(name, sourceToRun, breakFirstLine, print, module, baseUrl, compiledScript, origModule);
+}
+
+/**
+ * Wrapper of evalModuleEntryPoint
+ *
+ * This function wraps the compilation of the source code in a try-catch block.
+ * If the source code fails to be compiled, it will retry transpiling the source code
+ * with the TypeScript parser.
+ * @param {string} source The source code to evaluate
+ * @param {boolean} print If the result should be printed
+ * @returns {Promise} The module evaluation promise
+ */
+function evalTypeScriptModuleEntryPoint(source, print) {
+  if (print) {
+    throw new ERR_EVAL_ESM_CANNOT_PRINT();
+  }
+
+  RegExpPrototypeExec(/^/, ''); // Necessary to reset RegExp statics before user code runs.
+
+  return require('internal/modules/run_main').runEntryPointWithESMLoader(
+    async (loader) => {
+      const url = getEvalModuleUrl();
+      let moduleWrap;
+      try {
+        // Compile the module to check for syntax errors.
+        moduleWrap = loader.createModuleWrap(source, url);
+      } catch (originalError) {
+        // If it's not a SyntaxError, rethrow it.
+        if (!isError(originalError) || originalError.name !== 'SyntaxError') {
+          throw originalError;
+        }
+        let strippedSource;
+        try {
+          strippedSource = stripTypeScriptModuleTypes(source, url, false);
+          // If the moduleWrap was successfully created, execute the module job.
+          // outside the try-catch block to avoid catching runtime errors.
+          moduleWrap = loader.createModuleWrap(strippedSource, url);
+          // Emit the experimental warning after the code was successfully compiled.
+          emitExperimentalWarning('Type Stripping');
+        } catch (tsError) {
+          // If its not an error, or it's not an invalid typescript syntax error, rethrow it.
+          if (!isError(tsError) || tsError?.code !== 'ERR_INVALID_TYPESCRIPT_SYNTAX') {
+            throw tsError;
```

Comment:
What I found with syntax detection is that if you rethrow an exception in JavaScript, the first line (the source of the error) changes to Node internals rather than being the actual line of code the error originated from. The only way to prevent this was to do everything in C++. There might not be a test for this specifically for eval input; there should be, and there was for file input. I went down the same road of implementing the whole thing in JavaScript and stumbling onto this and then needing to redo it in C++ with the help of @targos and @joyeecheung.

Reply:
I mean it's eval does it really matter that we are changing the stacktrace?

Reply:
```
marcoippolito@marcos-MacBook-Pro node % node -e "throw new Error("yo")"                           
[eval]:1
throw new Error(yo)
                ^

ReferenceError: yo is not defined
    at [eval]:1:17
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:118:14
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:101:62)
    at evalScript (node:internal/process/execution:133:3)
    at node:internal/main/eval_string:51:3

Node.js v20.18.0
marcoippolito@marcos-MacBook-Pro node % ./node --experimental-strip-types -e "throw new Error("yo")"
[eval]:1
throw new Error(yo)
                ^

ReferenceError: yo is not defined
    at [eval]:1:17
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:478:12
    at [eval]-wrapper:6:24
    at runScriptInContext (node:internal/process/execution:476:60)
    at evalTypeScript (node:internal/process/execution:316:3)
    at node:internal/main/eval_string:71:3

Node.js v24.0.0-pre
```

I dont see any difference

Reply:
There is no difference:
```
marcoippolito@marcos-MacBook-Pro node % node -e "throw new Error('yo')"                                                
[eval]:1
throw new Error('yo')
^

Error: yo
    at [eval]:1:7
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:118:14
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:101:62)
    at evalScript (node:internal/process/execution:136:3)
    at node:internal/main/eval_string:55:3

Node.js v22.10.0
marcoippolito@marcos-MacBook-Pro node % ./node --experimental-strip-types -e "throw new Error('yo')"
[eval]:1
throw new Error('yo')
^

Error: yo
    at [eval]:1:7
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:478:12
    at [eval]-wrapper:6:24
    at runScriptInContext (node:internal/process/execution:476:60)
    at evalTypeScript (node:internal/process/execution:316:3)
    at node:internal/main/eval_string:71:3

Node.js v24.0.0-pre
```